cpp_polska/thread_local.md

```diff
@@ -0,0 +1,221 @@
+    ~Foo(1)
+
+Zgodnie z naszą intuicją, najpierw zostanie stworzona zmienna foo1, potem foo2, 
+następnie natrafiamy na koniec zakresu zmiennej foo2, więc zmienna ta zostaje 
+zniszczona. Na końcu, niszczy się foo1. Taki storage duration - od momentu zadeklarowania to końca zakresu - nosi nazwę `automatic storage duration`.
+
+### Static
+
+Kolejnym typem storage duration jest static storage duration.  Oznacza on, że zmienna zostanie stworzona w kolejności zadeklarowania , a usunięta na końcu 
+działania programu. Posiadają go zmienne zadeklarowane jako static, zmienne globalne oraz zmienne zadeklarowane jako extern.
+Poniżej ilustracja:
+
+    Foo foo3(3);
+
+    int main() {
+      Foo foo1(1); 
+      {
+        static Foo foo2(2);
+      }
+      return 0;
+    }
+
+Po uruchomieniu, na ekranie zobaczymy:
+
+    Foo(3)
+    Foo(1)
+    Foo(2)
+    ~Foo(1)
+    ~Foo(2)
+    ~Foo(3)
+
+Zmienne zostały stworzone w kolejności deklaracji. Jeśli chodzi o niszczenie - najpierw wywołał się destruktor foo1 (wyjście z zakresu), dalej zmienne są niszczone w kolejności odwrotnej do tworzenia, czyli wywołał się destruktor 
+zmiennej foo2 a potem zmiennej foo3. 
+
+Warto znać niuans dotyczący inicjalizacji zmiennych ze static storage duration. Rozważmy taki przykład: 
+
+    extern int i;
+    int j = i;
+    int i = 2;
+
+    int main() {
+      return j;
+    }
+
+W powyższym przykładzie, mogłoby się wydawać że zmienna `j` zostanie zainicjalizowana wartością 0. Jednak cały program zwraca wartość 2. 
+Spróbujmy napisać więc kontrprzykład, gdzie zmienna zostanie zainicjalizowana zerem: 
+
+    int foo() {
+      return 2;
+    }
+
+    extern int i;
+    int j = i;
+    int i = foo(); // = 2 before
+
+    int main() {
+      return j;
+    }
+
+Po tej niewielkiej zmianie, program zwraca wartość zero. Dlaczego tak się dzieje?
+Zerknijmy do standardu C++, dokładniej do [tego](http://eel.is/c++draft/basic.start.static#2) paragrafu:
+
+>All static initialization strongly happens before (...) any dynamic initialization.
+
+`int i = 2` jest inicjalizacją statyczną, natomiast `int i = foo()` jest inicjalizacją dynamiczną. Standard gwarantuje nam więc, że najpierw wykona się przypisanie do i.
+W drugim przypadku - obie inicjalizacje są dynamiczne, więc j będzie miało wartość 0 ([zero initialization](http://en.cppreference.com/w/cpp/language/zero_initialization)).
+
+### Thread
+Najmniej znanym typem storage duration, jest ten, który deklarujemy przy użyciu tytułowego słowa kluczowego **thread\_local**. 
+Aby zrozumieć różnicę pomiędzy static a thread\_local, posłużmy się przykładem.
+
+    void foo(int i) {
+      static Foo foo(i);
+    }
+
+    int main() {
+      foo(0);
+      std::thread th1(foo, 1);
+      std::thread th2(foo, 2);
+      th1.join();
+      th2.join();
+    }
+
+Na ekranie pojawi się:
+
+    Foo(0)
+    ~Foo(0)
+    
+Widzimy, że zostanie stworzona dokładnie jedna instancja klasy Foo. Zamieńmy teraz zmienną ze static na thread\_local:
```

Comment:
Na początku artykułu wspomniałeś, że jest to "pomocnicza **struktura**", może warto konsekwentnie się tego trzymać.